index.js

```diff
@@ -0,0 +1,88 @@
+'use strict'
+
+const fp = require('fastify-plugin')
+const auth = require('basic-auth')
+const FJS = require('fast-json-stringify')
+
+const serializeError = FJS({
+  type: 'object',
+  properties: {
+    statusCode: { type: 'number' },
+    error: { type: 'string' },
+    message: { type: 'string' }
+  }
+})
+
+function basicPlugin (fastify, opts, next) {
+  if (typeof opts.validate !== 'function') {
+    return next(new Error('Basic Auth: Missing validate function'))
+  }
+
+  const validate = opts.validate
+  fastify.decorate('basicAuth', basicAuth)
+  if (opts.disableHook !== true) {
+    fastify.addHook('onRequest', onBasicAuth)
+  }
+
+  next()
+
+  function onBasicAuth (req, res, next) {
+    var credentials = auth(req)
+    if (credentials == null) {
+      unauthorized()
+    } else {
+      var result = validate(credentials.name, credentials.pass, req, res, done)
+      if (result && typeof result.then === 'function') {
+        result.then(done, done)
+      }
+    }
+
+    function done (err) {
+      if (err !== undefined) {
+        unauthorized(err.message)
```

Comment:
I'm not 100% happy about this. I think we should have two different paths, one for errors and one for validation. Or we respect the fact that if `err.statusCode !== 401`  we use that code and error message.